Update and failing Match test

idkjs opened the pull request:
I was in here studying your modules and iterators and had to update it so here it is.

I can't get this test to pass and would greatly appreciate your feedback on how to get it to pass.


https://github.com/glennsl/bs-revamp/blob/270e7b0db581a3a9c8a461ccaa65c862feff7512/__tests__/revamp_test.re#L108-L113


I have tried a few differnt things that don't work.

https://github.com/idkjs/bs-revamp/blob/219e0ca1adbc6709a05a2b1e05576b40f13c608e/__tests__/revamp_test.re#L163-L177

```sh
  test("matches", () =>
    expect(
      Revamp.exec("a(n?)", "banana")
      |> List.fromSeq
      |> List.map(Revamp.Match.captures)
      |> List.map(Rebase.Array.fromList)
      // |> List.map(Js.Array.copy)
      //  |> Rebase.Array.copy
    )  /* matches has some extra properties attached, so to be able to compare it we need to convert it into a normal array */
    |> toEqual([
         [|Some("an"), Some("n")|],
         [|Some("an"), Some("n")|],
         [|Some("a"), Some("")|],
       ])
  );

```

<details>
<summary>## Test Report</summary>
```sh
$ npm run build && jest --coverage && npm run coverage

> @glennsl/bs-revamp@0.1.1 build
> bsb -make-world

Dependency on @glennsl/rebase
Dependency on @glennsl/bs-jest
Dependency on bs-benchmarkjs
Dependency Finished
 FAIL  lib/js/__tests__/revamp_test.js
  Compiled
    ✓ make - flags (5 ms)
    ✓ make - no flags (1 ms)
    ✕ test - raise Invalid_argument if given non-global regexp object (4 ms)
    ✕ test - raise Invalid_argument if given partially used global regexp object (2 ms)
    test
      ✓ match (1 ms)
      ✓ no match
    count
      ✓ match
      ✓ no match
    find
      ✓ match (1 ms)
      ✓ no match (2 ms)
    findIndex
      ✓ match (2 ms)
      ✓ no match (1 ms)
    replace
      ✓ match
      ✓ no match (1 ms)
    replaceByString
      ✓ match (1 ms)
      ✓ no match
    split
      ✓ match (1 ms)
      ✓ no match (1 ms)
  Match
    ✕ matches (7 ms)
    ✓ match (10 ms)
    ✓ captures
    ✓ captures - empty
    ✓ index (1 ms)
    ✓ input
  Uncompiled
    exec
      ✓ match (1 ms)
      ✓ no match
      ✓ flags
    matches
      ✓ match (1 ms)
      ✓ no match
      ✓ flags
    indices
      ✓ match
      ✓ no match (1 ms)
      ✓ flags
    captures
      ✓ match (1 ms)
      ✓ no match (1 ms)
      ✓ match - empty capture
      ✓ flags (1 ms)
    test
      ✓ match
      ✓ no match (1 ms)
      ✓ flags
    count
      ✓ match (1 ms)
      ✓ no match
      ✓ flags
    find
      ✓ match
      ✓ no match
      ✓ flags
    findIndex
      ✓ match (1 ms)
      ✓ no match
      ✓ findIndex
    replace
      ✓ match (1 ms)
      ✓ no match
      ✓ flags
    replaceByString
      ✓ match
      ✓ no match (1 ms)
      ✓ flags
    split
      ✓ match
      ✓ no match (1 ms)
      ✓ flags

  ● Compiled › test - raise Invalid_argument if given non-global regexp object

    expect(received).toThrow(expected)

    Expected substring: "{\"RE_EXN_ID\":\"Invalid_argument\",\"_1\":\"\"}"
    Received value:     {"Error": [Error], "RE_EXN_ID": "Invalid_argument", "_1": "Invalid RegEx, global == false (This should never happen, file a bug!)"}

      at affirm (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:153:38)
      at Object.<anonymous> (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:238:11)

  ● Compiled › test - raise Invalid_argument if given partially used global regexp object

    expect(received).toThrow(expected)

    Expected substring: "{\"RE_EXN_ID\":\"Invalid_argument\",\"_1\":\"\"}"
    Received value:     {"Error": [Error], "RE_EXN_ID": "Invalid_argument", "_1": "Invalid RegEx, lastIndex should be 0, is 2 (This should never happen, file a bug!)"}

      at affirm (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:153:38)
      at Object.<anonymous> (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:238:11)

  ● Match › matches

    expect(received).toEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 0

    @@ -1,18 +1,15 @@
      Object {
        "hd": Array [
    -     "an",
          "n",
        ],
        "tl": Object {
          "hd": Array [
    -       "an",
            "n",
          ],
          "tl": Object {
            "hd": Array [
    -         "a",
              "",
            ],
            "tl": 0,
          },
        },

      at affirm (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:82:38)
      at Object.<anonymous> (node_modules/@glennsl/bs-jest/lib/js/src/jest.js:238:11)

-----------|---------|----------|---------|---------|-------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s
-----------|---------|----------|---------|---------|-------------------
All files  |   99.04 |      100 |   97.44 |   99.04 |
 Revamp.js |   99.04 |      100 |   97.44 |   99.04 | 52
-----------|---------|----------|---------|---------|-------------------
Test Suites: 1 failed, 1 total
Tests:       3 failed, 55 passed, 58 total
Snapshots:   0 total
Time:        2.539 s
Ran all test suites.
```
</details>

Thanks for sharing your work.

glennsl commented on src/Revamp.rei:
```diff
@@ -31,7 +31,7 @@ module Compiled: {
   let replace: (t, string => string, string) => string;
   let replaceByString: (t, string, string) => string; /* TODO: better name? */
-  let split: (t, string) => array(string); /* TODO: should be list string? */
+  let split: (t, Js.String.t) => array(option(Js.String.t)); /* TODO: should be list string? */
```
Why `Js.String.t`? It should just be an alias of `string` anyway.

This change to `option` will need a mention in the change log (at the bottom of the README)

idkjs replied:
As you noted, `Js.String.t` and `string` behave the same way. I have reverted it. To answer the question as to why, I don't know. I was just running through the errors until it compiled. I imagine I got an error saying the interfaces did not matched and wholesale copied the type signature in the console to the interface file. I did not take care, as, as you say, we were kinda just making it work. Also, I did not appreciate that one was an alias for the other.

glennsl commented on yarn.lock:
```diff
@@ -0,0 +1,3985 @@
+# THIS IS AN AUTOGENERATED FILE. DO NOT EDIT THIS FILE DIRECTLY.
+# yarn lockfile v1
```
I don't use `yarn`, so this shouldn't be checked in. And `package-lock.json` should be updated. I've already done that in https://github.com/idkjs/bs-revamp/pull/1 though

idkjs replied:
Thank you for this detailed guidance, sir. I had checked in the yarn.lock because I could not get this to build with npm for some reason. Initially, it was the old bs-platform so i bailed out and used yarn. I tried to rebuild with npm afterwards but still did not work so...yarn. Thanks for fixing it. 

Thanks for pointing this out and the other comments you made. 

Your comment about formatting above is one I have heard before. How do you keep your editor from changing the formatting? Usually, I am in vscode. Should I take out any formatting configuration or is the solution something else? On "serious" commits, I find myself going through the output and fixing the unnecessary formatting changes where there isnt a `.prettierrc` or linter in the project.

glennsl replied:
You should be able to just set `"editor.formatOnSave": false` in your `settings.json`. IMO format on save ought to be disabled globally by default, and only enabled per-project. That'll avoid any accidental formatting.

glennsl reviewed (commented):
Thanks for this! The formatting changes make it very difficult to see what's actually changed though. And if this was an actively maintained project I'd probably reject it because it makes the code significantly less readable. In general it's not a good idea to use automatic formatting when contributing to other projects.

It looks like the tests need some cleaning up though. Otherwise, with PR I suggested, it's just the few minor issues below.

idkjs commented:
Last issue is that coveralls is throwing this error:
```sh
/Volumes/SSD/Github/bs-revamp/node_modules/coveralls/bin/coveralls.js:19
      throw err;
      ^
Bad response: 422 {"message":"Couldn't find a repository matching this job.","error":true}
```

First time ever looking at coveralls. I will try to chase this down.

<details>
<summary>Full Coveralls Output</summary>
```sh

> @glennsl/bs-revamp@0.1.1 test
> npm run build && jest --coverage && npm run coverage


> @glennsl/bs-revamp@0.1.1 build
> bsb -make-world

Dependency on @glennsl/rebase
Dependency on @glennsl/bs-jest
Dependency on bs-benchmarkjs
Dependency Finished
-----------|---------|----------|---------|---------|-------------------
File       | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s 
-----------|---------|----------|---------|---------|-------------------
All files  |     100 |      100 |     100 |     100 |                   
 Revamp.js |     100 |      100 |     100 |     100 |                   
-----------|---------|----------|---------|---------|-------------------

> @glennsl/bs-revamp@0.1.1 coverage
> nyc report --temp-directory=coverage --reporter=text-lcov | coveralls --verbose

[debug] "2021-03-14T17:54:36.893Z"  'TN:\n' +
  'SF:lib/js/src/Revamp.js\n' +
  'FN:6,_captures\n' +
  'FN:7,(anonymous_1)\n' +
  'FN:16,_match\n' +
  'FN:20,_flagToString\n' +
  'FN:32,_assertValid\n' +
  'FN:51,index\n' +
  'FN:55,input\n' +
  'FN:66,make\n' +
  'FN:68,(anonymous_8)\n' +
  'FN:74,exec\n' +
  'FN:76,(anonymous_10)\n' +
  'FN:86,(anonymous_11)\n' +
  'FN:91,(anonymous_12)\n' +
  'FN:96,matches\n' +
  'FN:98,(anonymous_14)\n' +
  'FN:103,indices\n' +
  'FN:105,(anonymous_16)\n' +
  'FN:106,(anonymous_17)\n' +
  'FN:116,captures\n' +
  'FN:118,(anonymous_19)\n' +
  'FN:123,test\n' +
  'FN:130,find\n' +
  'FN:140,findIndex\n' +
  'FN:154,count\n' +
  'FN:158,replace\n' +
  'FN:162,replaceByString\n' +
  'FN:166,split\n' +
  'FN:185,exec$1\n' +
  'FN:190,matches$1\n' +
  'FN:195,indices$1\n' +
  'FN:200,captures$1\n' +
  'FN:205,test$1\n' +
  'FN:210,find$1\n' +
  'FN:215,findIndex$1\n' +
  'FN:220,count$1\n' +
  'FN:225,replace$1\n' +
  'FN:230,replaceByString$1\n' +
  'FN:235,split$1\n' +
  'FNF:38\n' +
  'FNH:38\n' +
  'FNDA:11,_captures\n' +
  'FNDA:27,(anonymous_1)\n' +
  'FNDA:24,_match\n' +
  'FNDA:14,_flagToString\n' +
  'FNDA:42,_assertValid\n' +
  'FNDA:3,index\n' +
  'FNDA:3,input\n' +
  'FNDA:43,make\n' +
  'FNDA:14,(anonymous_8)\n' +
  'FNDA:25,exec\n' +
  'FNDA:66,(anonymous_10)\n' +
  'FNDA:41,(anonymous_11)\n' +
  'FNDA:25,(anonymous_12)\n' +
  'FNDA:5,matches\n' +
  'FNDA:5,(anonymous_14)\n' +
  'FNDA:3,indices\n' +
  'FNDA:3,(anonymous_16)\n' +
  'FNDA:4,(anonymous_17)\n' +
  'FNDA:4,captures\n' +
  'FNDA:4,(anonymous_19)\n' +
  'FNDA:7,test\n' +
  'FNDA:5,find\n' +
  'FNDA:5,findIndex\n' +
  'FNDA:5,count\n' +
  'FNDA:2,replace\n' +
  'FNDA:2,replaceByString\n' +
  'FNDA:2,split\n' +
  'FNDA:8,exec$1\n' +
  'FNDA:3,matches$1\n' +
  'FNDA:3,indices$1\n' +
  'FNDA:4,captures$1\n' +
  'FNDA:3,test$1\n' +
  'FNDA:3,find$1\n' +
  'FNDA:3,findIndex$1\n' +
  'FNDA:3,count$1\n' +
  'FNDA:3,replace$1\n' +
  'FNDA:3,replaceByString$1\n' +
  'FNDA:3,split$1\n' +
  'DA:3,1\n' +
  'DA:4,1\n' +
  'DA:7,11\n' +
  'DA:8,27\n' +
  'DA:9,2\n' +
  'DA:11,25\n' +
  'DA:17,24\n' +
  'DA:21,14\n' +
  'DA:23,12\n' +
  'DA:25,1\n' +
  'DA:27,1\n' +
  'DA:33,42\n' +
  'DA:34,1\n' +
  'DA:35,1\n' +
  'DA:41,41\n' +
  'DA:42,40\n' +
  'DA:44,1\n' +
  'DA:52,3\n' +
  'DA:56,3\n' +
  'DA:59,1\n' +
  'DA:67,43\n' +
  'DA:68,43\n' +
  'DA:69,14\n' +
  'DA:71,43\n' +
  'DA:75,25\n' +
  'DA:76,25\n' +
  'DA:77,66\n' +
  'DA:78,66\n' +
  'DA:79,66\n' +
  'DA:80,25\n' +
  'DA:82,41\n' +
  'DA:83,41\n' +
  'DA:84,41\n' +
  'DA:87,41\n' +
  'DA:91,25\n' +
  'DA:92,25\n' +
  'DA:97,5\n' +
  'DA:98,5\n' +
  'DA:99,5\n' +
  'DA:104,3\n' +
  'DA:105,3\n' +
  'DA:106,3\n' +
  'DA:107,4\n' +
  'DA:108,4\n' +
  'DA:117,4\n' +
  'DA:118,4\n' +
  'DA:119,4\n' +
  'DA:124,7\n' +
  'DA:125,5\n' +
  'DA:126,5\n' +
  'DA:127,5\n' +
  'DA:131,5\n' +
  'DA:132,5\n' +
  'DA:133,5\n' +
  'DA:134,3\n' +
  'DA:135,3\n' +
  'DA:141,5\n' +
  'DA:142,5\n' +
  'DA:143,5\n' +
  'DA:144,2\n' +
  'DA:146,3\n' +
  'DA:147,3\n' +
  'DA:148,3\n' +
  'DA:155,5\n' +
  'DA:159,2\n' +
  'DA:163,2\n' +
  'DA:167,2\n' +
  'DA:170,1\n' +
  'DA:186,8\n' +
  'DA:187,8\n' +
  'DA:191,3\n' +
  'DA:192,3\n' +
  'DA:196,3\n' +
  'DA:197,3\n' +
  'DA:201,4\n' +
  'DA:202,4\n' +
  'DA:206,3\n' +
  'DA:207,3\n' +
  'DA:211,3\n' +
  'DA:212,3\n' +
  'DA:216,3\n' +
  'DA:217,3\n' +
  'DA:221,3\n' +
  'DA:222,3\n' +
  'DA:226,3\n' +
  'DA:227,3\n' +
  'DA:231,3\n' +
  'DA:232,3\n' +
  'DA:236,3\n' +
  'DA:237,3\n' +
  'DA:240,1\n' +
  'DA:241,1\n' +
  'DA:242,1\n' +
  'DA:243,1\n' +
  'DA:244,1\n' +
  'DA:245,1\n' +
  'DA:246,1\n' +
  'DA:247,1\n' +
  'DA:248,1\n' +
  'DA:249,1\n' +
  'DA:250,1\n' +
  'DA:251,1\n' +
  'DA:252,1\n' +
  'LF:103\n' +
  'LH:103\n' +
  'BRDA:8,0,0,2\n' +
  'BRDA:8,0,1,25\n' +
  'BRDA:21,1,0,12\n' +
  'BRDA:21,1,1,1\n' +
  'BRDA:21,1,2,1\n' +
  'BRDA:33,2,0,1\n' +
  'BRDA:33,2,1,41\n' +
  'BRDA:41,3,0,40\n' +
  'BRDA:41,3,1,1\n' +
  'BRDA:67,4,0,40\n' +
  'BRDA:67,4,1,3\n' +
  'BRDA:79,5,0,25\n' +
  'BRDA:79,5,1,41\n' +
  'BRDA:133,6,0,3\n' +
  'BRDA:133,6,1,2\n' +
  'BRDA:143,7,0,2\n' +
  'BRDA:143,7,1,3\n' +
  'BRDA:186,8,0,1\n' +
  'BRDA:186,8,1,7\n' +
  'BRDA:191,9,0,1\n' +
  'BRDA:191,9,1,2\n' +
  'BRDA:196,10,0,1\n' +
  'BRDA:196,10,1,2\n' +
  'BRDA:201,11,0,1\n' +
  'BRDA:201,11,1,3\n' +
  'BRDA:206,12,0,1\n' +
  'BRDA:206,12,1,2\n' +
  'BRDA:211,13,0,1\n' +
  'BRDA:211,13,1,2\n' +
  'BRDA:216,14,0,1\n' +
  'BRDA:216,14,1,2\n' +
  'BRDA:221,15,0,1\n' +
  'BRDA:221,15,1,2\n' +
  'BRDA:226,16,0,1\n' +
  'BRDA:226,16,1,2\n' +
  'BRDA:231,17,0,1\n' +
  'BRDA:231,17,1,2\n' +
  'BRDA:236,18,0,1\n' +
  'BRDA:236,18,1,2\n' +
  'BRF:39\n' +
  'BRH:39\n' +
  'end_of_record\n'
[debug] "2021-03-14T17:54:36.896Z"  'user options undefined'
[debug] "2021-03-14T17:54:36.899Z"  'No valid .coveralls.yml file found'
[debug] "2021-03-14T17:54:36.958Z"  {
  run_at: '2021-03-14T17:54:36.897Z',
  git: {
    head: {
      id: '3f63a02df00baa00f2bc88babc7ac8d71449d94a',
      committer_name: 'Alain Armand',
      committer_email: 'aarmand.inbox@gmail.com',
      message: 'revert Js.String.t to string  and Rebase.List.t to list per https://github.com/glennsl/bs-revamp/pull/1#discussion_r593901426',
      author_name: 'Alain Armand',
      author_email: 'aarmand.inbox@gmail.com'
    },
    branch: 'reason',
    remotes: [ [Object], [Object] ]
  }
}
[debug] "2021-03-14T17:54:36.960Z"  'in: ' ''
[info] "2021-03-14T17:54:36.963Z"  'sending this to coveralls.io: ' `{"source_files":[{"name":"lib/js/src/Revamp.js","source":"'use strict';\\n\\nvar Rebase = require(\\"@glennsl/rebase/lib/js/src/Rebase.js\\");\\nvar Caml_option = require(\\"bs-platform/lib/js/caml_option.js\\");\\n\\nfunction _captures(result) {\\n  return Rebase.$$Option.getOrRaise(Rebase.List.tail(Rebase.List.map((function (prim) {\\n                        if (prim == null) {\\n                          return ;\\n                        } else {\\n                          return Caml_option.some(prim);\\n                        }\\n                      }), Rebase.List.fromArray(result))));\\n}\\n\\nfunction _match(result) {\\n  return Rebase.$$Array.unsafeGetUnchecked(0, result);\\n}\\n\\nfunction _flagToString(param) {\\n  switch (param) {\\n    case /* IgnoreCase */0 :\\n        return \\"i\\";\\n    case /* MultiLine */1 :\\n        return \\"m\\";\\n    case /* Unicode */2 :\\n        return \\"u\\";\\n    \\n  }\\n}\\n\\nfunction _assertValid(re) {\\n  if (re.lastIndex !== 0) {\\n    var lastIndex = re.lastIndex;\\n    throw {\\n          RE_EXN_ID: \\"Invalid_argument\\",\\n          _1: \\"Invalid RegEx, lastIndex should be 0, is \\" + lastIndex + \\" (This should never happen, file a bug!)\\",\\n          Error: new Error()\\n        };\\n  }\\n  if (re.global === true) {\\n    return ;\\n  }\\n  throw {\\n        RE_EXN_ID: \\"Invalid_argument\\",\\n        _1: \\"Invalid RegEx, global == false (This should never happen, file a bug!)\\",\\n        Error: new Error()\\n      };\\n}\\n\\nfunction index(prim) {\\n  return prim.index;\\n}\\n\\nfunction input(prim) {\\n  return prim.input;\\n}\\n\\nvar Match = {\\n  match_: _match,\\n  captures: _captures,\\n  index: index,\\n  input: input\\n};\\n\\nfunction make(flagsOpt, pattern) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  var flags$1 = Rebase.List.reduce((function (acc, flag) {\\n          return acc + flag;\\n        }), \\"g\\", Rebase.List.map(_flagToString, flags));\\n  return new RegExp(pattern, flags$1);\\n}\\n\\nfunction exec(re, input) {\\n  _assertValid(re);\\n  var next = function (start, param) {\\n    re.lastIndex = start;\\n    var result = re.exec(input);\\n    if (result === null) {\\n      return /* Nil */0;\\n    }\\n    var nextIndex = re.lastIndex;\\n    re.lastIndex = 0;\\n    return /* Cons */{\\n            _0: result,\\n            _1: (function (param) {\\n                return next(nextIndex, param);\\n              })\\n          };\\n  };\\n  return function (param) {\\n    return next(0, param);\\n  };\\n}\\n\\nfunction matches(re, input) {\\n  var partial_arg = exec(re, input);\\n  return function (param) {\\n    return Rebase.Seq.map(_match, partial_arg, param);\\n  };\\n}\\n\\nfunction indices(re, input) {\\n  var partial_arg = exec(re, input);\\n  return function (param) {\\n    return Rebase.Seq.map((function (result) {\\n                  var index = result.index;\\n                  return [\\n                          index,\\n                          index + _match(result).length | 0\\n                        ];\\n                }), partial_arg, param);\\n  };\\n}\\n\\nfunction captures(re, input) {\\n  var partial_arg = exec(re, input);\\n  return function (param) {\\n    return Rebase.Seq.map(_captures, partial_arg, param);\\n  };\\n}\\n\\nfunction test(re, input) {\\n  _assertValid(re);\\n  var res = re.test(input);\\n  re.lastIndex = 0;\\n  return res;\\n}\\n\\nfunction find(re, input) {\\n  _assertValid(re);\\n  var result = re.exec(input);\\n  if (result !== null) {\\n    re.lastIndex = 0;\\n    return Caml_option.some(_match(result));\\n  }\\n  \\n}\\n\\nfunction findIndex(re, input) {\\n  _assertValid(re);\\n  var result = re.exec(input);\\n  if (result === null) {\\n    return ;\\n  }\\n  re.lastIndex = 0;\\n  var index = result.index;\\n  return [\\n          index,\\n          index + _match(result).length | 0\\n        ];\\n}\\n\\nfunction count(re, input) {\\n  return Rebase.Seq.count(exec(re, input));\\n}\\n\\nfunction replace(re, f, input) {\\n  return input.replace(re, f);\\n}\\n\\nfunction replaceByString(re, replacement, input) {\\n  return input.replace(re, replacement);\\n}\\n\\nfunction split(re, input) {\\n  return input.split(re);\\n}\\n\\nvar Compiled = {\\n  make: make,\\n  exec: exec,\\n  matches: matches,\\n  indices: indices,\\n  captures: captures,\\n  test: test,\\n  find: find,\\n  findIndex: findIndex,\\n  count: count,\\n  replace: replace,\\n  replaceByString: replaceByString,\\n  split: split\\n};\\n\\nfunction exec$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return exec(make(flags, pattern), input);\\n}\\n\\nfunction matches$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return matches(make(flags, pattern), input);\\n}\\n\\nfunction indices$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return indices(make(flags, pattern), input);\\n}\\n\\nfunction captures$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return captures(make(flags, pattern), input);\\n}\\n\\nfunction test$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return test(make(flags, pattern), input);\\n}\\n\\nfunction find$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return find(make(flags, pattern), input);\\n}\\n\\nfunction findIndex$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return findIndex(make(flags, pattern), input);\\n}\\n\\nfunction count$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return count(make(flags, pattern), input);\\n}\\n\\nfunction replace$1(pattern, flagsOpt, f, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return input.replace(make(flags, pattern), f);\\n}\\n\\nfunction replaceByString$1(pattern, flagsOpt, replacement, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return input.replace(make(flags, pattern), replacement);\\n}\\n\\nfunction split$1(pattern, flagsOpt, input) {\\n  var flags = flagsOpt !== undefined ? flagsOpt : /* [] */0;\\n  return input.split(make(flags, pattern));\\n}\\n\\nexports.Match = Match;\\nexports.Compiled = Compiled;\\nexports.exec = exec$1;\\nexports.matches = matches$1;\\nexports.indices = indices$1;\\nexports.captures = captures$1;\\nexports.test = test$1;\\nexports.find = find$1;\\nexports.findIndex = findIndex$1;\\nexports.count = count$1;\\nexports.replace = replace$1;\\nexports.replaceByString = replaceByString$1;\\nexports.split = split$1;\\n/* No side effect */\\n","coverage":[null,null,1,1,null,null,11,27,2,null,25,null,null,null,null,null,24,null,null,null,14,null,12,null,1,null,1,null,null,null,null,null,42,1,1,null,null,null,null,null,41,40,null,1,null,null,null,null,null,null,null,3,null,null,null,3,null,null,1,null,null,null,null,null,null,null,43,43,14,null,43,null,null,null,25,25,66,66,66,25,null,41,41,41,null,null,41,null,null,null,25,25,null,null,null,null,5,5,5,null,null,null,null,3,3,3,4,4,null,null,null,null,null,null,null,null,4,4,4,null,null,null,null,7,5,5,5,null,null,null,5,5,5,3,3,null,null,null,null,null,5,5,5,2,null,3,3,3,null,null,null,null,null,null,5,null,null,null,2,null,null,null,2,null,null,null,2,null,null,1,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,8,8,null,null,null,3,3,null,null,null,3,3,null,null,null,4,4,null,null,null,3,3,null,null,null,3,3,null,null,null,3,3,null,null,null,3,3,null,null,null,3,3,null,null,null,3,3,null,null,null,3,3,null,null,1,1,1,1,1,1,1,1,1,1,1,1,1,null,null],"branches":[8,0,0,2,8,0,1,25,21,1,0,12,21,1,1,1,21,1,2,1,33,2,0,1,33,2,1,41,41,3,0,40,41,3,1,1,67,4,0,40,67,4,1,3,79,5,0,25,79,5,1,41,133,6,0,3,133,6,1,2,143,7,0,2,143,7,1,3,186,8,0,1,186,8,1,7,191,9,0,1,191,9,1,2,196,10,0,1,196,10,1,2,201,11,0,1,201,11,1,3,206,12,0,1,206,12,1,2,211,13,0,1,211,13,1,2,216,14,0,1,216,14,1,2,221,15,0,1,221,15,1,2,226,16,0,1,226,16,1,2,231,17,0,1,231,17,1,2,236,18,0,1,236,18,1,2]}],"git":{"head":{"id":"3f63a02df00baa00f2bc88babc7ac8d71449d94a","committer_name":"Alain Armand","committer_email":"aarmand.inbox@gmail.com","message":"revert Js.String.t to string  and Rebase.List.t to list per https://github.com/glennsl/bs-revamp/pull/1#discussion_r593901426","author_name":"Alain Armand","author_email":"aarmand.inbox@gmail.com"},"branch":"reason","remotes":[{"name":"origin","url":"https://github.com/idkjs/bs-revamp.git"},{"name":"upstream","url":"https://github.com/glennsl/bs-revamp.git"}]},"run_at":"2021-03-14T17:54:36.897Z"}`
```
</details>

glennsl commented:
> Last issue is that coveralls is throwing this error

Don't worry about it. Coveralls is just for coverage reporting on CI, so it doesn't matter if it fails outside CI.

glennsl commented on __bench__/iterators.re:
```diff
@@ -35,7 +35,7 @@ module Seq = {
 
   let matches = (input, re) =>
     re |> exec(input)
-       |> map(result => Array.unsafeGetUnchecked(0, result |> Js.Re.matches));
+       |> map(result => Array.unsafeGetUnchecked(0, result |> Js.Re.captures));
```
This is misleadingly named now that it's changed.

idkjs replied:
updated in [`eed6d34` (#1)](https://github.com/glennsl/bs-revamp/pull/1/commits/eed6d34ca793a32b3d96cb5ca1f80f6ecd92886c)

idkjs replied:
Any guidance you could provid on getting my editor to respect the formatting in this repo would be appreciated. Thank you.

glennsl replied:
Waat you did with the `rebase` PR seems to have worked great! I don;t expect you to revert the formatting here if it's too much work though.

glennsl commented on __tests__/revamp_test.re:
```diff
@@ -1,300 +1,445 @@
-    let result = 
-      {js|
+  describe("Compiled", () => {
+    open Expect;
+    open Rebase;
```
What's the reason for moving this locally?

idkjs replied:
I was trying isolate the error in the tests below. I though maybe there were some conflicts which there were not. I have removed the local opens. Thanks.

glennsl commented on __tests__/revamp_test.re:
```diff
@@ -1,300 +1,445 @@
   describe("split", () => {
-    test("match", () => 
-      expect(Revamp.split(" |!", "bang bang bananabatman!")) |> toEqual(([|"bang", "bang", "bananabatman", ""|])));
+    // open Rebase;
```
odd randomly commented out line

idkjs replied:
fixed

glennsl commented on __tests__/revamp_test.re:
```diff
@@ -1,300 +1,445 @@
+             Some(""),
+           |])
+      );
+      // expect(Revamp.Compiled.split(re, "bang bang bananabatman!")) |> toEqual(([|"bang", "bang", "bananabatman", ""|])));
```
Another commented out line

idkjs replied:
Fixed

glennsl commented on src/Revamp.rei:
```diff
@@ -31,7 +28,7 @@ module Compiled: {
   let replace: (t, string => string, string) => string;
   let replaceByString: (t, string, string) => string; /* TODO: better name? */
-  let split: (t, string) => array(string); /* TODO: should be list string? */
+  let split: (t, string) => array(option(string)); /* TODO: should be list string? */
```
I think the API here would be better if the implementation was changed instead to filter out the `None`s. That would break with JS behaviour of course, but I think it's very rare that this would be useful. There's an example of the behaviour here: https://github.com/rescript-lang/rescript-compiler/issues/3120#issuecomment-433637078

idkjs replied:
I appreciate your allowing me to take a shot at this. The most recent commit is my attempt to update the api per your guidance. I am not at all sure that this is what you had in mind. It seems to work and the tests are passing.

glennsl replied:
See https://github.com/glennsl/bs-revamp/pull/1#discussion_r594552674

glennsl reviewed (commented):
It's still very hard to see what's actually changed in the tests because of the formatting changes, but I've managed to find a few issues there still. As well as some other suggestions:

glennsl commented on src/Revamp.re:
```diff
@@ -2,7 +2,7 @@ open Rebase;
 [@bs.set] external _setLastIndex : (Js.Re.t, int) => unit = "lastIndex";
 [@bs.send.pipe : string] external _replace : (Js.Re.t, string => string) => string = "replace";
-
+[@bs.send.pipe: string] external _splitByRe: Js.Re.t => array(string) = "split";
```
This is unsound, so all this'll do is reintroduce the unsoundness that was fixed in BuckleScript. I'll show what I meant below.

glennsl commented on src/Revamp.re:
```diff
@@ -131,8 +129,7 @@ module Compiled = {
 
-  let split = (re, input) =>
-    Js.String.splitByRe(re, input);
+  let split = (re, input) => _splitByRe(re, input);
```
This is what I meant:

```suggestion
  let split = (re, input) => Js.String.splitByRe(re, input)->Belt.Array.keepMap(x => x);
```

Might look odd, but if you look at signature of `leepMap` the function is expected to return an option, where `Some(x)` means the item is kept and `None` means it's dropped. Which makes it really easy for us to filter out the `None`s and unwrapping the `option`s in one go.

glennsl commented:
Great work. Thanks so much @idkjs!

idkjs commented:
@glennsl Thank you, sir. And thank you for taking the time to teach. Greatly appreciated.
